### DISCUSSION TITLE
Emby: hardware acceleration problem because the permissions

### DISCUSSION THREAD

**Discussion by LeidenSpain** [tteck/Proxmox#747] [Help Needed] [answered]
Hi,

I have installed emby using your script, and everything is perfect, except that the graphics acceleration doesn't work for me. I have an integrated intel.

I have read a lot (a lot) about this problem, I have seen that it also happens with plex, etc...
I have consulted in the emby forums, and in the end we have come to the conclusion that it is a problem with the permissions of the users.
https://emby.media/community/index.php?/topic/113382-hw-acceleration-in-lxc-on-proxmox/

This is my lxc
```
# Emby LXC
#### https%3A//tteck.github.io/Proxmox/
#<a href='https%3A//ko-fi.com/D1D7EP4GF'><img src='https%3A//img.shields.io/badge/%E2%98%95-Buy me a coffee-red' /></a>
arch: amd64
cores: 2
features: nesting=1
hostname: emby
memory: 2048
mp0: /media/USB1,mp=/media/USB1
net0: name=eth0,bridge=vmbr0,gw=192.168.3.1,hwaddr=9E:D3:B9:2C:85:D5,ip=192.168.3.99/24,type=veth
onboot: 0
ostype: ubuntu
rootfs: local-lvm:vm-199-disk-0,size=8G
swap: 512
lxc.cgroup2.devices.allow: c 226:0 rwm
lxc.cgroup2.devices.allow: c 226:128 rwm
lxc.cgroup2.devices.allow: c 29:0 rwm
lxc.mount.entry: /dev/fb0 dev/fb0 none bind,optional,create=file
lxc.mount.entry: /dev/dri dev/dri none bind,optional,create=dir
lxc.mount.entry: /dev/dri/renderD128 dev/renderD128 none bind,optional,create=file

```

These are the users to access the igpu
```
HOST
getent group video | cut -d: -f3
44
getent group render | cut -d: -f3
103


LXC
getent group video | cut -d: -f3
44
getent group render | cut -d: -f3
107
```

My subuid and subgid files (both are the same)
```
root:100000:65536
root:44:1
root:103:1
root:104:1
```

Let's see if you could help me clarify the reason, because I no longer understand anything.

Regards

**Comment by LeidenSpain** [comment]
Hi,

I have seen that in the machine the folder group renderD128 is not render.
```
root@server:~# ls -l /dev/dri/
total 0
drwxr-xr-x 2 root root         80 Dec 27 20:25 by-path
crw-rw---- 1 root video  226,   0 Dec 27 20:25 card0
crw-rw---- 1 root render 226, 128 Dec 27 20:25 renderD128
```


And now, from the CT:
```
root@emby:~# ls -l /dev/dri/
total 0
drwxr-xr-x 2 root root         80 Dec 27 19:25 by-path
crw-rw---- 1 root video  226,   0 Dec 27 19:25 card0
crw-rw---- 1 root syslog 226, 128 Dec 27 19:25 renderD128
```

Could this be the problem for not being able to use hardware acceleration?

Regards

**Reply by tteck** [reply]
I know that mount point got updated on Dec. 10th 
https://github.com/tteck/Proxmox/pull/829/files

**Comment by LeidenSpain** [comment]
Hi,

I don't have that much knowledge, but I think we are talking about different problems.

The one you mention was that the wrong mount point (/dev/renderD128).
The problem now is that renderD128 does not belong to the render group, but to the syslog

My last test CT is created this week.

Regards

**Comment by LeidenSpain** [comment]
Hi,

Yesterday I finally got the tip that has allowed the emby LXC to work perfectly.
Indeed, I have solved the problem by modifying the permissions of the /dev/dri/ folder

`chmod 666 /dev/dri/*` 

This change is removed when the host machine is rebooted, so I will have to add it to cron so that it is modified at each start.

Hi,

 

Yesterday I finally got the tip that has allowed the emby LXC to work perfectly.
Indeed, I have solved the problem by modifying the permissions of the /dev/dri/ folder

chmod 666 /dev/dri/*
 

This change is removed when the host machine is rebooted, so I will have to add it to cron so that it is modified at each start.

So it seems clear that there is something wrong with the permissions.
What I don't understand is that no one else has this problem. Could it be that they don't use hardware acceleration?

Regards

**Reply by RealJustMe** [reply]
Hey,

I had the same problem with the permissions and added 
`lxc.hook.pre-start: sh -c "chown 0:107 /dev/dri/renderD128"`
to my "/etc/pve/lxc/xxx.conf"

Survives every reboot :)

**Reply by LeidenSpain** [reply, marked as answer]
> Hey,
> 
> I had the same problem with the permissions and added `lxc.hook.pre-start: sh -c "chown 0:107 /dev/dri/renderD128"` to my "/etc/pve/lxc/xxx.conf"
> 
> Survives every reboot :)

great solution, much better than modifying the host.
I think it would be very interesting if the script added this line, or gave the option to do it or not

Thanks

**Reply by nigiriemoji** [reply]
> Hey,
> 
> I had the same problem with the permissions and added `lxc.hook.pre-start: sh -c "chown 0:107 /dev/dri/renderD128"` to my "/etc/pve/lxc/xxx.conf"
> 
> Survives every reboot :)

I'm having trouble getting this to work. Through some testing, running `chmod 666 /dev/dri/*` in my jellyfin lxc container gets the transcoding working.
My XXX.conf file looks like this:

```
## Jellyfin LXC
#  ### https%3A//tteck.github.io/Proxmox/
#  <a href='https%3A//ko-fi.com/D1D7EP4GF'><img src='https%3A//img.shields.io/badge/%E2%98%95-Buy me a coffee-red' /></a>
arch: amd64
cores: 4
features: nesting=1
hostname: jellyfin
memory: 2048
mp0: /mnt/14TB,mp=/mnt/14TB
mp1: /mnt/8TB,mp=/mnt/8TB
net0: name=eth0,bridge=vmbr0,hwaddr=XX:XX:XX:XX:XX,ip=dhcp,type=veth
onboot: 1
ostype: ubuntu
rootfs: local-zfs:subvol-102-disk-0,size=32G
swap: 512
tags: proxmox-helper-scripts
lxc.cgroup2.devices.allow: a
lxc.cap.drop:
lxc.cgroup2.devices.allow: c 226:0 rwm
lxc.cgroup2.devices.allow: c 226:128 rwm
lxc.cgroup2.devices.allow: c 29:0 rwm
lxc.mount.entry: /dev/fb0 dev/fb0 none bind,optional,create=file
lxc.mount.entry: /dev/dri dev/dri none bind,optional,create=dir
lxc.mount.entry: /dev/dri/renderD128 dev/dri/renderD128 none bind,optional,create=file
lxc.hook.pre-start: sh -c "chown 0:107 /dev/dri/renderD128"
```

**Reply by RealJustMe** [reply]
Be sure that 107 is the correct ID for the render group, as it may be not 107 for Jellyfin.
In your LXC run: `getent group render | cut -d: -f3`

**Reply by mattwoods5** [reply]
While this resolves the problem for the LXC, it is changing the group on the host. Is there a way we can map the render group to be consistent across the host and container?

**Reply by RealJustMe** [reply]
Yea, I realized it a few days ago. I was tinkering with it but only had success with a unprivileged LXC and ID Mapping. 
I use `lxc.hook.pre-start: sh -c "chmod 666 /dev/dri/*"` for now but keep working on it.

**Comment by clartek** [comment]
Just wanted to say that `chmod 666 /dev/dri/*` worked for me, and I now see my integrated GPU in the Emby transcoding settings. The hw acceleration script in the misc folder doesn't seem to work fully.

**Comment by birkhofflee** [comment]
TL;DR:

I had to use the solution provided by @RealJustMe to have HW acceleration work.

First, `getent group render | cut -d: -f3` says `104` on my machine.

Then add the line

```
lxc.hook.pre-start: sh -c "chown 0:104 /dev/dri/renderD128"
```

in Emby LXC configuration. Now the transcoding page should show the hardware options correctly. FYI, you can choose the Advanced option in the Transcoding settings page to see if Emby actually detects HW acceleration:

![CleanShot 2024-05-14 at 17 03 45@2x](https://github.com/tteck/Proxmox/assets/8148095/c41ca3d2-2af2-4b93-b0c6-e005b8ee6b64)

after confirming you better put it back to *Yes* to use the default options.

@tteck I think this could be included in the setup script to make it work automatically. :-) Cheers
